### PR TITLE
[model/predict] Add support for different types of action outputs 

### PR DESCRIPTION
**Pull request by ANarayan**
This PR adds support processing action outputs of different types. The current implementation expects that all action outputs are of type tensor. Here, we expand support for outputs of type dict. This PR also adds a new utils function (`merge_objects`) which merges different types of objects.

### CODE REVIEW

File: src/emmental/model.py [hunk 1]
```diff
@@ -403,9 +403,15 @@ def forward(  # type: ignore
                 and self.action_outputs[task_name] is not None
             ):
                 for action_name, output_index in self.action_outputs[task_name]:
-                    out_dict[task_name][f"{action_name}_{output_index}"] = (
-                        output_dict[action_name][output_index].cpu().detach().numpy()
-                    )
+                    action_output = output_dict[action_name][output_index]
+                    if isinstance(action_output, dict):
```
Comment (senwu) [inline src/emmental/model.py:407]: Can we move this if-else to `move_to_device`?

File: src/emmental/model.py [hunk 2]
```diff
@@ -526,21 +532,23 @@ def predict(
                 if return_action_outputs and out_bdict:
                     for task_name in out_bdict.keys():
                         for action_name in out_bdict[task_name].keys():
-                            out_dict[task_name][action_name].extend(
-                                out_bdict[task_name][action_name]
-                            )
+                            if out_dict[task_name][action_name] == []:
```
Comment (senwu) [inline src/emmental/model.py:535]: Can we move this to `merge_objects`?

File: src/emmental/utils/utils.py [hunk 3]
```diff
@@ -158,6 +158,49 @@ def move_to_device(
...
 
+def merge_objects(obj_1: Any, obj_2: Any) -> Any:
+    """Merge two objects of the same type.
+
+    Given two objects of the same type and structure, merges the second object
+    into the first object.
+
+    Args:
+      obj_1: first object.
+      obj_2: seecond object to be merged into the first object.
```
Comment (senwu) [inline src/emmental/utils/utils.py:169]: typo?

File: src/emmental/utils/utils.py [hunk 4]
```diff
@@ -158,6 +158,49 @@ def move_to_device(
...
+
```
Comment (senwu) [inline src/emmental/utils/utils.py:170]: Can you add more description about this function here?

File: src/emmental/utils/utils.py [hunk 5]
```diff
@@ -158,6 +158,49 @@ def move_to_device(
...
+    Returns:
+      an object reflecting the merged output of the two inputs.
+    """
+    if isinstance(obj_1, torch.Tensor):
```
Comment (senwu) [inline src/emmental/utils/utils.py:174]: We need check the two objects are the same type right?

File: src/emmental/utils/utils.py [hunk 6]
```diff
@@ -158,6 +158,49 @@ def move_to_device(
...
+        return torch.cat([obj_1, obj_2])
+    elif isinstance(obj_1, dict):
+        if not obj_1:
+            return obj_2
+        elif not obj_2:
```
Comment (senwu) [inline src/emmental/utils/utils.py:179]: What's this?

File: src/emmental/utils/utils.py [hunk 7]
```diff
@@ -158,6 +158,49 @@ def move_to_device(
...
+            return obj_1
+        else:
+            for key, value in obj_1.items():
+                if isinstance(value, torch.Tensor):
```
Comment (senwu) [inline src/emmental/utils/utils.py:183]: also include `np.array`?

File: src/emmental/utils/utils.py [hunk 8]
```diff
@@ -158,6 +158,49 @@ def move_to_device(
...
+                    obj_1[key] = np.concatenate((value, obj_2[key]))
+                elif isinstance(value, list):
+                    obj_1[key].extend(obj_2[key])
+            return obj_1
+    elif isinstance(obj_1, np.ndarray):
+        return np.concatenate((obj_1, obj_2))
+    elif isinstance(obj_1, list):
+        obj_1.extend(obj_2)
+        return obj_1
+    elif isinstance(obj_1, tuple):
```
Comment (senwu) [inline src/emmental/utils/utils.py:195]: Tuple might have more than 2 objects.

File: tests/utils/test_utils.py [hunk 9]
```diff
@@ -68,6 +69,27 @@ def test_move_to_device(caplog):
     assert move_to_device((torch.tensor([1, 2]), torch.tensor([3, 4])), -1)
 
 
+def test_merge_objects(caplog):
+    """Unit test of merge_objects."""
+    caplog.set_level(logging.INFO)
+
+    assert torch.equal(
+        merge_objects(torch.Tensor([1, 2]), torch.Tensor([2, 3])),
+        torch.Tensor([1, 2, 2, 3]),
```
Comment (senwu) [inline tests/utils/test_utils.py:78]: This should be `torch.Tensor[[1, 2], [2, 3]]`? We want to merge based on the first dim.

File: tests/utils/test_utils.py [hunk 10]
```diff
@@ -68,6 +69,27 @@ def test_move_to_device(caplog):
...
+    )
+    assert np.array_equal(
+        merge_objects(np.array([1, 2]), np.array([2, 3])), np.array([1, 2, 2, 3])
```
Comment (senwu) [inline tests/utils/test_utils.py:81]: Save here?

File: tests/utils/test_utils.py [hunk 11]
```diff
@@ -68,6 +69,27 @@ def test_move_to_device(caplog):
...
+    )
+    assert merge_objects({"a": [1, 2]}, {"a": [2, 3]}) == {"a": [1, 2, 2, 3]}
+    assert merge_objects({"a": [1, 2]}, {}) == {"a": [1, 2]}
+    assert merge_objects({}, {"a": [1, 2]}) == {"a": [1, 2]}
+    assert merge_objects(([2, 4], [3, 4]), ([3, 4], [4, 5])) == (
+        [2, 4, 3, 4],
+        [3, 4, 4, 5],
+    )
+    assert merge_objects([1, 2, 3], [2, 3, 4]) == [1, 2, 3, 2, 3, 4]
+
```
Comment (senwu) [inline tests/utils/test_utils.py:91]: can we have more use cases about `np.array`, `torch.Tensor`?

**Review by senwu — COMMENTED** [review]
Left some comments

File: src/emmental/utils/utils.py (hunk 7 above)
Comment (ANarayan) [reply src/emmental/utils/utils.py:183]: @senwu isn't the type of a np.array `<class 'numpy.ndarray'>`

File: src/emmental/utils/utils.py (hunk 6 above)
Comment (ANarayan) [reply src/emmental/utils/utils.py:179]: case when either of the two dict objects are empty

File: src/emmental/model.py (hunk 2 above)
Comment (ANarayan) [reply src/emmental/model.py:535]: problem is that `out_dict[task_name][action_name]` is always going to be a list, but out_bdict[task_name][action_name] can be any type. Thus, the `merge_objects` function will just raise a type error because of type mistmatch

Comment (senwu) [reply src/emmental/model.py:531]: this should be gone right?

File: tests/e2e/test_e2e.py [hunk 12]
```diff
@@ -297,6 +408,11 @@ def forward(self, input):
         for idx in range(len(test2_pred["outputs"]["task2"]["_input__data"]))
     ]
 
+    assert len(test4_pred["outputs"]["task3"]["input1_t3_out"]["image_pil"]) == 10
+    assert isinstance(
+        test4_pred["outputs"]["task3"]["input1_t3_out"]["image_pil"], np.ndarray
```
Comment (senwu) [inline tests/e2e/test_e2e.py:413]: Can we check the shape?

**Review by senwu — COMMENTED** [review]
Left few comments

File: tests/e2e/test_e2e.py (hunk 12 above)
Comment (ANarayan) [reply tests/e2e/test_e2e.py:413]: for https://github.com/senwu/emmental/pull/113#discussion_r740705320, we can't get rid of it because `out_dict[task_name][action_name]` and `out_bdict[task_name][action_name]` may not be of the same type for the empty check to be handled by `merge_objects`
